src/Http/WebUtilities/src/MultipartPipeReader.cs

```diff
@@ -0,0 +1,353 @@
+                }
+
+                colon = line.IndexOf(ColonDelimiter);
+
+                if (colon == -1)
+                {
+                    throw new InvalidDataException($"Invalid header line: {line.ToString()}");
+                }
+                else
+                {
+                    if (line.Length > lengthLimit)
+                    {
+                        throw new InvalidDataException($"Line length limit {lengthLimit} exceeded.");
+                    }
+                    key = line.Slice(0, colon);
+                    value = line.Slice(colon + ColonDelimiter.Length);
+                }
+
+                lengthLimit -= line.Length;
+                var decodedKey = GetDecodedString(key);
+                var decodedValue = GetDecodedString(value).Trim();
+
+                AppendAndVerify(ref accumulator, decodedKey, decodedValue);
+            }
+            return false;
+        }
+
+        // For multi-segment parsing of a read only sequence
+        private bool ParseHeadersSlow(
+            ref ReadOnlySequence<byte> buffer,
+            ref KeyValueAccumulator accumulator,
+            ref long headersLength,
+            bool isFinalBlock)
+        {
+            var sequenceReader = new SequenceReader<byte>(buffer);
+            var position = sequenceReader.Position;
+            var consumedBytes = default(long);
+
+            while (!sequenceReader.End)
+            {
+                if (!sequenceReader.TryReadTo(out ReadOnlySequence<byte> line, CrlfDelimiter))
+                {
+                    if (!isFinalBlock)
+                    {
+                        // Don't buffer indefinately
+                        if (headersLength + consumedBytes > HeadersLengthLimit)
+                        {
+                            throw new InvalidDataException($"Line length limit {HeadersLengthLimit - headersLength} exceeded.");
+                        }
+                        break;
+                    }
+
+                    // This must be the final key=value pair
+                    line = buffer.Slice(sequenceReader.Position);
+                    consumedBytes = sequenceReader.Consumed;
+                    sequenceReader.Advance(line.Length);
+                }
+
+                if (consumedBytes + headersLength > HeadersLengthLimit)
+                {
+                    throw new InvalidDataException($"Multipart headers length limit {HeadersLengthLimit} exceeded.");
+                }
+
+                if (line.IsSingleSegment)
+                {
+                    var didFinishParsing = ParseHeadersFast(line.FirstSpan, ref accumulator, isFinalBlock: true, HeadersLengthLimit - consumedBytes - headersLength, out var segmentConsumed);
+                    Debug.Assert(segmentConsumed == line.FirstSpan.Length);
+                    consumedBytes = sequenceReader.Consumed;
+                    position = sequenceReader.Position;
+                    if (didFinishParsing)
```

Comment:
This shouldn't happen (unless the line was empty?)

Reply:
Oops you are right. It also currently returns false even when the length is empty and it doesn't enter the while loop.